### PR TITLE
Messing with component mobs further

### PR DESCRIPTION
**Pull request by MadmanMartian**
- Moves the NPC Component holder to living-level, from living/component and human level. Makes for easier NPC-NPC communication
- Adds Comsig_attackedby so they're informed they are in fact being bullied.
- Adds Comsig_GetDefZone and Comsig_SetDefZone for when you're attacking a human and want to say 'yeah mob, kick him in the dick'

- [x] Fix their melee combat. Currently runtimes as they're not attacking anything specific.
https://pastebin.com/0cEym42Y
- [x] Look into using Astar pathing rather than the crap default byond stuff, so we can add tracking exceptions (Won't try to path to something that's behind a window or door, for instance)
- [x] Learn proper component standards.
 ## Current Problems
~~- When attacking. Should they be near a backpack and their item is on their back, they will be caught in a loop of constantly switching the item to and fro their hands.~~
 ~~- When moving on AStar, it is not returning an Astar list of nodes.~~

### CODE REVIEW

**Comment by DamianX** [comment]
upon what time

**Comment by DamianX** [comment]
is this done?

**Comment by MadmanMartian** [comment]
Not buggy, but sucks and is full of debug.

Need to add an astar subcontroller, but that'll have to be in another PR

**Comment by MadmanMartian** [comment]
Should be good to go. Astar subprocess'll need to be in another PR

File: code/_onclick/click.dm [hunk 1]
```diff
@@ -118,7 +118,13 @@
 			item_attack_delay = held_item.attack_delay
 			var/resolved = held_item.preattack(A, src, 1, params)
 			if(!resolved)
-				resolved = A.attackby(held_item, src, params)
+				if(ismob(A) && modifiers["def_zone"])
+					var/mob/M = A
+					var/def_zone
+					def_zone = modifiers["def_zone"]
+					resolved = M.attackby(held_item,src,def_zone = def_zone)
```
Comment (DamianX) [inline code/_onclick/click.dm]: Is there a reason you're not passing `params` here? I legit don't know

File: code/modules/components/ai/def_zone_handler.dm [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/datum/component/ai/targetting_handler/RecieveAndReturnSignal(var/message_type, var/list/args)
+	if(message_type == COMSIG_GETDEFZONE)
+		var/mob/living/target = args["target"]
+		var/damagetype = args["damage_type"]
+		ASSERT(target && istype(target))
```
Comment (DamianX) [inline code/modules/components/ai/def_zone_handler.dm]: `istype(target)` is sufficient

File: code/modules/components/ai/def_zone_handler.dm [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		return EvaluateTarget(target, damagetype)
+
+/datum/component/ai/targetting_handler/proc/EvaluateTarget(var/mob/living/target, var/damagetype) //Center mass.
+	return LIMB_CHEST
+
+/datum/component/ai/targetting_handler/dumb/EvaluateTarget(var/mob/living/target, var/damagetype) //Random
+	var/list/potential_targets = list(
```
Comment (DamianX) [inline code/modules/components/ai/def_zone_handler.dm]: This should be a `var/static/list` so it's not recreated each time this runs

File: code/modules/components/ai/def_zone_handler.dm [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		LIMB_RIGHT_HAND,
+		LIMB_LEFT_LEG,
+		LIMB_RIGHT_LEG,
+		LIMB_LEFT_FOOT,
+		LIMB_RIGHT_FOOT,
+		TARGET_MOUTH)
+	return pick(potential_targets)
+
+/datum/component/ai/targetting_handler/smart/EvaluateTarget(var/mob/living/target, var/damagetype) //Goes for the part with the least armor
+	var/list/potential_target = list(
```
Comment (DamianX) [inline code/modules/components/ai/def_zone_handler.dm]: This should be a `var/static/list` so it's not recreated each time this runs

File: code/modules/components/ai/human/human_target_finder.dm [hunk 5]
```diff
@@ -18,4 +18,14 @@
...
-	return o
\ No newline at end of file
+	return o
+
+/datum/component/ai/target_finder/human/RecieveSignal(var/message_type, var/list/args)
+	..()
+	if(message_type == COMSIG_ATTACKEDBY) //YOU HAVE JUST MADE AN ENEMY FOR LIFE
+		var/assailant = args["assailant"]
+		var/damage_done = args["damage"]
+		if(B.friends.Find(assailant) && damage_done > 15) //Intent to kill!
```
Comment (DamianX) [inline code/modules/components/ai/human/human_target_finder.dm]: Just check for `damage_done > 15`, `Find` is superfluous

File: code/modules/components/ai/human/human_target_finder.dm [hunk 6]
```diff
@@ -18,4 +18,14 @@
...
+			B.friends.Remove(assailant)
+		if(damage_done > 2 && !B.enemies.Find(assailant))
```
Comment (DamianX) [inline code/modules/components/ai/human/human_target_finder.dm]: You should replace `!Find` and `Add` with `B.enemies |= assailant`

File: code/modules/components/ai/human/human_target_finder.dm [hunk 7]
```diff
@@ -18,4 +18,14 @@
...
\ No newline at end of file
+	return o
+
+/datum/component/ai/target_finder/human/RecieveSignal(var/message_type, var/list/args)
+	..()
+	if(message_type == COMSIG_ATTACKEDBY) //YOU HAVE JUST MADE AN ENEMY FOR LIFE
+		var/assailant = args["assailant"]
+		var/damage_done = args["damage"]
+		if(damage_done > 15) //Intent to kill!
+			B.friends &= assailant
```
Comment (MadmanMartian) [inline code/modules/components/ai/human/human_target_finder.dm]: @DamianX Does this work? I'm not familiar with these operators regarding lists.

File: code/modules/components/ai/human/human_target_finder.dm [hunk 8]
```diff
@@ -18,4 +18,14 @@
...
+		if(damage_done > 2)
+			B.enemies |= assailant
```
Comment (MadmanMartian) [inline code/modules/components/ai/human/human_target_finder.dm]: Same with this.

File: code/modules/components/ai/human/human_target_finder.dm (hunk 7 above)
Comment (DamianX) [reply code/modules/components/ai/human/human_target_finder.dm]: You originally had `B.friends.Remove(assailant)`, which was fine.
The equivalent would be `-=` but I did not request this change. I just said the `Find()` was not needed.
So no, this does not work.

File: code/modules/components/ai/human/human_target_finder.dm (hunk 8 above)
Comment (DamianX) [reply code/modules/components/ai/human/human_target_finder.dm]: This is good.

**Review by DamianX — APPROVED** [review]

**Review by DamianX — CHANGES_REQUESTED** [review]
Fuck, I meant to request changes.

**Review by DamianX — APPROVED** [review]

**Comment by MadmanMartian** [comment]
This should be good to go. The humans won't exactly be very imposing until I can add an astar subprocess, rather than it running off of life

File: code/modules/components/debugger.dm [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+/datum/component/debugger/RecieveSignal(var/message_type, var/list/args)
+	if(spam)
+		to_chat(world, "===================================")
+		to_chat(world, "[container.holder] received [message_type], args are")
+		for(var/i in args)
+			to_chat(world, "[i]: [args[i]]")
+		to_chat(world, "<a HREF='?src=\ref[src];pause=1'>\[Press here to stop and start the spam\]</a>")
+
+/datum/component/debugger/Topic(href, href_list)
+	.=.()
```
Comment (DamianX) [inline code/modules/components/debugger.dm]: oh no, this should be ..(), yeah?

**Review by DamianX — CHANGES_REQUESTED** [review]

**Pull request merged**